File: solutions/523.cpp
class Solution {
public:
    bool checkSubarraySum(vector<int>& nums, int k) {
        int n = nums.size();
        if (n < 2) return false;
        for (int i = 0; i < n; ++i) {
            nums[i] %= k;
        }
        
        unordered_set<int> st;
        int prevSum = 0;
        int currentSum = 0;
        for (int i = 0; i < n; ++i) {
            if (i == 0) {
                currentSum += nums[i];
            }
            if (i >= 1) {
                currentSum += nums[i];
                st.insert(prevSum % k);
                if (st.find(currentSum % k) != st.end()) return true;
                prevSum += nums[i - 1];
            }
        }
        return false;
    }
};

File: rust_solutions/523.rs
impl Solution {
    pub fn check_subarray_sum(nums: Vec<i32>, k: i32) -> bool {
        let mut hash = std::collections::HashSet::new();
        let mut current = 0;
        let mut previous = 0;

        let n = nums.len();
        if n < 2 {
            return false;
        }
        for i in 0..n {
            current += nums[i];
            if i > 0 {
                hash.insert(previous % k);
                let key = current % k;
                if hash.contains(&key) {
                    return true;
                }
                previous += nums[i - 1];
            }
        }
        return false;
    }
}